### PR TITLE
feat: add firewall policies

### LINKED ISSUE #387: Implementation of firewall policy + FQDN Filtering - data classification [BUILD]
1. As a Platform Admin, I need to ensure that no traffic can flow across data classification even if a firewall rule is defined to allow that traffic
2. As a Platform Admin, I need to allow traffic for approved FQDN
3. As an Application Developer, I need to allow traffic to FQDN

**Implementation details:**

The designed solution leverages Hierarchical Firewall Policies.

The network POC has policy: phac-aspc-generic-fwpol

Currently visible under folder:
Non-IaC\Network POC\landing-zone\clients\phac-aspc\standard\applications-infrastructure

**It requires the definition of :**

- Firewall policy
- Firewall policy rule (some in client-t2, pot. some in t3)
- Firewall policy association
- Determine in what repo and package should each resources resides (T1,2,3 ).

**Scope:**
Includes naming convention + pot. labels.

**Acceptance Criteria:**

- Firewall policy
- Firewall policy rule (some in client-t2, pot. some in t3)
- Firewall policy association

### PR DESCRIPTION
**Pull request by alaincormier-ssc**
closes #387 
add firewall policy and dependencies
provide example for a tier3 rule

`core-landing-zone`
- update org custom role permissions

`client-setup`
- update client namespaces service account permissions

`client-landing-zone`
- add three levels of firewall policies and rules
- add cloud nat/router
- update vpc settings

`client-project-setup`
- update tier3-sa permissions

credits to @lucstjean-ssc @davelanglois-ssc

### CODE REVIEW

**Review by cartyc — APPROVED** [review]
Nice work Alain! Great to see the FQDN rules land :)

**Review by fmichaelobrien — APPROVED** [review]
LGTM

File: solutions/client-landing-zone/client-folder/standard/applications-infrastructure/firewall-policy/rules/defaults.yaml [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+# 2147483645 default ingress rule ipv6 Ingress IPv6 ranges: ::/0 all Goto next
+# 2147483646 default egress rule Egress IPv4 ranges: 0.0.0.0/0 all Goto next
+# 2147483647 default ingress rule Ingress IPv4 ranges: 0.0.0.0/0 all Goto next
+#########
+# Exclude communication with private IP ranges, leaving only Internet traffic to be inspected (EGRESS)
+apiVersion: compute.cnrm.cloud.google.com/v1beta1
+kind: ComputeFirewallPolicyRule
+metadata:
+  name: client-name-fwpol-exclude-private-ip-ranges-egress-fwr # kpt-set: ${client-name}-fwpol-exclude-private-ip-ranges-egress-fwr
+  namespace: client-name-networking # kpt-set: ${client-name}-networking
```
Comment (davelanglois-ssc) [inline solutions/client-landing-zone/client-folder/standard/applications-infrastructure/firewall-policy/rules/defaults.yaml:27]: please add a depends-on annotations for all compute firewallpolicyrule

**Review by davelanglois-ssc — CHANGES_REQUESTED** [review]

**Review by Michael-McAuley-SSC — DISMISSED** [review]
/lgtm

File: solutions/client-landing-zone/client-folder/standard/applications-infrastructure/host-project/network/nat.yaml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#########
+# Cloud NAT northamerica-northeast1
+apiVersion: compute.cnrm.cloud.google.com/v1beta1
+kind: ComputeRouterNAT
+metadata:
+  name: nane1-nat
```
[the remark this answers is not in the document]
Comment (alaincormier-ssc) [reply solutions/client-landing-zone/client-folder/standard/applications-infrastructure/host-project/network/nat.yaml]: updated

File: solutions/client-landing-zone/client-folder/firewall-policy/policy.yaml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+# 2147483646 default egress rule Egress IPv4 ranges: 0.0.0.0/0 all Goto next
+# 2147483647 default ingress rule Ingress IPv4 ranges: 0.0.0.0/0 all Goto next
+#########
+# Client Compute Firewall Policy on folder client-folder
+apiVersion: compute.cnrm.cloud.google.com/v1beta1
+kind: ComputeFirewallPolicy
+metadata:
+  name: client-name-client-folder-fwpol # kpt-set: ${client-name}-client-folder-fwpol
+  namespace: client-name-networking # kpt-set: ${client-name}-networking
+spec:
```
Comment (davelanglois-ssc) [inline solutions/client-landing-zone/client-folder/firewall-policy/policy.yaml:27]: missing depends-on

File: solutions/client-landing-zone/client-folder/standard/applications-infrastructure/firewall-policy/policy.yaml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+# 2147483645 default ingress rule ipv6 Ingress IPv6 ranges: ::/0 all Goto next
+# 2147483646 default egress rule Egress IPv4 ranges: 0.0.0.0/0 all Goto next
+# 2147483647 default ingress rule Ingress IPv4 ranges: 0.0.0.0/0 all Goto next
+#########
+# Client Compute Firewall Policy on folder standard.applications-infrastructure
+apiVersion: compute.cnrm.cloud.google.com/v1beta1
+kind: ComputeFirewallPolicy
+metadata:
+  name: client-name-standard-applications-infrastructure-fwpol # kpt-set: ${client-name}-standard-applications-infrastructure-fwpol
+  namespace: client-name-networking # kpt-set: ${client-name}-networking
```
Comment (davelanglois-ssc) [inline solutions/client-landing-zone/client-folder/standard/applications-infrastructure/firewall-policy/policy.yaml:26]: missing depends on

File: solutions/client-landing-zone/client-folder/standard/firewall-policy/policy.yaml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+# 2147483645 default ingress rule ipv6 Ingress IPv6 ranges: ::/0 all Goto next
+# 2147483646 default egress rule Egress IPv4 ranges: 0.0.0.0/0 all Goto next
+# 2147483647 default ingress rule Ingress IPv4 ranges: 0.0.0.0/0 all Goto next
+#########
+# Client Compute Firewall Policy on folder standard
+apiVersion: compute.cnrm.cloud.google.com/v1beta1
+kind: ComputeFirewallPolicy
+metadata:
+  name: client-name-standard-fwpol # kpt-set: ${client-name}-standard-fwpol
+  namespace: client-name-networking # kpt-set: ${client-name}-networking
```
Comment (davelanglois-ssc) [inline solutions/client-landing-zone/client-folder/standard/firewall-policy/policy.yaml:26]: missing depends on

**Review by davelanglois-ssc — CHANGES_REQUESTED** [review]

File: solutions/client-landing-zone/client-folder/standard/firewall-policy/policy.yaml (hunk 5 above)
Comment (alaincormier-ssc) [reply solutions/client-landing-zone/client-folder/standard/firewall-policy/policy.yaml:26]: added

File: solutions/client-landing-zone/client-folder/standard/applications-infrastructure/firewall-policy/policy.yaml (hunk 4 above)
Comment (alaincormier-ssc) [reply solutions/client-landing-zone/client-folder/standard/applications-infrastructure/firewall-policy/policy.yaml:26]: added

File: solutions/client-landing-zone/client-folder/firewall-policy/policy.yaml (hunk 3 above)
Comment (alaincormier-ssc) [reply solutions/client-landing-zone/client-folder/firewall-policy/policy.yaml:27]: can't seem to add a depends-on for a folder that was created in another root-sync
leaving out for now, we'll have to investigate

**Review by davelanglois-ssc — APPROVED** [review]

**Pull request merged**